core/amber/src/main/scala/edu/uci/ics/amber/engine/common/ambermessage/WorkerMessage.scala

```diff
@@ -2,76 +2,71 @@ package edu.uci.ics.amber.engine.common.ambermessage
 
-  final case class ExecutionCompleted() extends WorkflowMessage
+  final case class ExecutionCompleted() extends ControlEvent
 
-  final case class ExecutionPaused() extends WorkflowMessage
+  final case class ExecutionPaused() extends ControlEvent
 
-  final case class AssignBreakpoint(breakpoint: LocalBreakpoint) extends WorkflowMessage
+  final case class AssignBreakpoint(breakpoint: LocalBreakpoint) extends ControlEvent
 
-  final case class QueryTriggeredBreakpoints() extends WorkflowMessage
+  final case class QueryTriggeredBreakpoints() extends ControlEvent
 
-  final case class QueryBreakpoint(id: String) extends WorkflowMessage
+  final case class QueryBreakpoint(id: String) extends ControlEvent
 
-  final case class ReportState(workerState: WorkerState.Value) extends WorkflowMessage
+  final case class ReportState(workerState: WorkerState.Value) extends ControlEvent
 
-  final case class ReportStatistics(workerStatistics: WorkerStatistics) extends WorkflowMessage
+  final case class ReportStatistics(workerStatistics: WorkerStatistics) extends ControlEvent
 
-  final case class ReportOutputResult(results: List[ITuple]) extends WorkflowMessage
+  final case class ReportOutputResult(results: List[ITuple]) extends ControlEvent
 
-  final case class RemoveBreakpoint(id: String) extends WorkflowMessage
+  final case class RemoveBreakpoint(id: String) extends ControlEvent
 
-  final case class ReportedTriggeredBreakpoints(bps: Array[LocalBreakpoint]) extends WorkflowMessage
+  final case class ReportedTriggeredBreakpoints(bps: Array[LocalBreakpoint]) extends ControlEvent
 
-  final case class ReportedQueriedBreakpoint(bp: LocalBreakpoint) extends WorkflowMessage
+  final case class ReportedQueriedBreakpoint(bp: LocalBreakpoint) extends ControlEvent
 
-  final case class ReportFailure(exception: Exception) extends WorkflowMessage
+  final case class ReportFailure(exception: Exception) extends ControlEvent
 
-  final case class ReportUpstreamExhausted(tag: LayerTag) extends WorkflowMessage
+  final case class ReportUpstreamExhausted(tag: LayerTag) extends ControlEvent
 
   final case class ReportWorkerPartialCompleted(worker: WorkerTag, layer: LayerTag)
-      extends WorkflowMessage
+      extends ControlEvent
 
-  final case class CheckRecovery() extends WorkflowMessage
+  final case class CheckRecovery() extends ControlEvent
 
   final case class ReportCurrentProcessingTuple(workerID: ActorPath, tuple: ITuple)
-      extends WorkflowMessage
+      extends ControlEvent
+
+  final case class Reset(core: Any, recoveryInformation: Seq[(Long, Long)]) extends ControlEvent
 
-  final case class Reset(core: Any, recoveryInformation: Seq[(Long, Long)]) extends WorkflowMessage
+  final case class EndOfUpstream() extends DataEvent
 
-  final case class DataMessage(sequenceNumber: Long, payload: Array[ITuple])
-      extends WorkflowMessage {
+  final case class DataPayload(payload: Array[ITuple]) extends DataEvent {
```

Comment:
let's call it dataFrame

Reply:
let's standardize what we are calling different things:
1. DataFrame -> Array[ITuple]
2.  EndOfUpstreamMarker ??
3. SenderTuplePair -> (Int, ITuple)
4. DataPayload -> two types - DataFrame and EndOfUpstream
5. Similarly ControlPayload
6. WorkflowDataMessage -> DataPayload  + from + seqNum
7. WorkflowControlMessage -> similarly
8. WorkflowMessage -> two types
9. NetworkMessage -> WorkflowMessage + uuid. I don't know why NetworkMessage has to be different. Let;s chat